client/src/common/fileOperations.ts

```diff
@@ -249,6 +285,22 @@ export class DidRenameFilesFeature extends NotificationFileOperationFeature<{ ol
 		);
 	}
 
+	public register(data: RegistrationData<proto.FileOperationRegistrationOptions>): void {
+		if (!this._willListener) {
+			this._willListener = code.workspace.onWillRenameFiles(this.willRename, this);
+		}
+		super.register(data);
+	}
+
+	private willRename(e: code.FileWillRenameEvent): void {
+		// Calling filter will force the matching logic to run. For any item
```

Comment:
We shouldn't make the will rename wait on this. Instead we should cache promises and make the actual did wait for the promise to resolve. This simply slows down the VS Code operation for no need.

Reply:
Similar to above - can there be concurrent requests? Will we need to track multiple promises, or can we just track one? (if it's called again, we could always replace the promise with another chained into the end - that would avoid any issues if there were concurrent requests though could result in the earlier requests waiting for the later requests depending on timing).

Reply:
I would argue that we can track the last assuming that it reflects the current state in the file system.

Reply:
> We shouldn't make the will rename wait on this. Instead we should cache promises and make the actual did wait for the promise to resolve. This simply slows down the VS Code operation for no need.

I had a go at this, but I'm seeing failures and I don't think this is possible. If we don't _block_ VS Code from performing the operation while we read the types, the files may be moved/deleted _before_ our async work to get the file types has completed and we'll fail to get them.

For rename, we could perhaps work around it by using the _new_ locations, but that wouldn't be consistent/possible with delete.

(I've done the other changes to clear cache + change to Map though).